nodeup/pkg/model/awsebscsidriver.go

```diff
@@ -1,58 +0,0 @@
-/*
-Copyright 2020 The Kubernetes Authors.
-
-Licensed under the Apache License, Version 2.0 (the "License");
-you may not use this file except in compliance with the License.
-You may obtain a copy of the License at
-
-    http://www.apache.org/licenses/LICENSE-2.0
-
-Unless required by applicable law or agreed to in writing, software
-distributed under the License is distributed on an "AS IS" BASIS,
-WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-See the License for the specific language governing permissions and
-limitations under the License.
-*/
-
-package model
-
-import (
-	"k8s.io/kops/upup/pkg/fi"
-)
-
-// AWSEBSCSIDriverBuilder writes AWSEBSCSIDriver's assets
-type AWSEBSCSIDriverBuilder struct {
-	*NodeupModelContext
-}
-
-var _ fi.ModelBuilder = &AWSEBSCSIDriverBuilder{}
-
-// Build is responsible for configuring the EBS CSI Driver stuff
-func (b *AWSEBSCSIDriverBuilder) Build(c *fi.ModelBuilderContext) error {
-	csi := b.Cluster.Spec.CloudConfig.AWSEBSCSIDriver
-
-	if csi == nil || !fi.BoolValue(csi.Enabled) {
-		return nil
-	}
-
-	// Pulling CSI driver image
-	image := "k8s.gcr.io/provider-aws/aws-ebs-csi-driver:" + *csi.Version
```

Comment:
There's value in doing the warmpull. We should keep this, but figure out why we're dereferencing nil here.

Reply:
I disagree in this case. This should not have been done at all for anything other than CNI and proxy, which affect startup.
A generic way, with container images assets, would be the way to go forward.

Reply:
If you have pending workloads with volumes this will speed up pod startup. So while it does not contribute to node startup, it contributed to pod scheduling time.

Reply:
The nil dereferencing works because it assumes cloudup always set this value. That is the case now, but I wouldn't exactly call this very defensive.

Reply:
This is a general argument for any pre-pulled image. Realistically, the pull time for these images is very small.
As I said, I don't disagree with the idea, if done correctly. I am not ok with using it in this case and having extra logic in NodeUp for a special case.

Reply:
Note quite. There are three images pulled in succession, and one of them sometimes takes up to 20 seconds.
A random example:
```
FirstSeen              Count    From      Type     Reason      Message
<nil>                  <none>   <none>    Normal   Scheduled   Successfully assigned kube-system/ebs-csi-node-9sbqq to ip-172-20-98-90.eu-central-1.compute.internal
2021-06-03T17:34:17Z   1        kubelet   Normal   Pulling     Pulling image "k8s.gcr.io/provider-aws/aws-ebs-csi-driver:v1.0.0"
2021-06-03T17:34:20Z   1        kubelet   Normal   Pulled      Successfully pulled image "k8s.gcr.io/provider-aws/aws-ebs-csi-driver:v1.0.0" in 3.187918027s
2021-06-03T17:34:21Z   1        kubelet   Normal   Created     Created container ebs-plugin
2021-06-03T17:34:21Z   1        kubelet   Normal   Started     Started container ebs-plugin
2021-06-03T17:34:21Z   1        kubelet   Normal   Pulling     Pulling image "k8s.gcr.io/sig-storage/csi-node-driver-registrar:v2.1.0"
2021-06-03T17:34:34Z   1        kubelet   Normal   Pulled      Successfully pulled image "k8s.gcr.io/sig-storage/csi-node-driver-registrar:v2.1.0" in 12.506681986s
2021-06-03T17:34:34Z   1        kubelet   Normal   Created     Created container node-driver-registrar
2021-06-03T17:34:34Z   1        kubelet   Normal   Started     Started container node-driver-registrar
2021-06-03T17:34:34Z   1        kubelet   Normal   Pulling     Pulling image "k8s.gcr.io/sig-storage/livenessprobe:v2.2.0"
2021-06-03T17:34:35Z   1        kubelet   Normal   Pulled      Successfully pulled image "k8s.gcr.io/sig-storage/livenessprobe:v2.2.0" in 1.094773878s
2021-06-03T17:34:35Z   1        kubelet   Normal   Created     Created container liveness-probe
2021-06-03T17:34:36Z   1        kubelet   Normal   Started     Started container liveness-probe
```

Reply:
Probably we can debate this forever. I doubt any of us will change its mind. Hopefully this will change in the near future to something more maintainable.